### PR TITLE
Willroman/create json

### PR DESCRIPTION
**Pull request by willroman**
Fixes https://github.com/aspnet/AspNetCore/issues/17779

Replacement pull request for Willroman/Builder #22 (differenced against aspnet/SignalR-Client-Cpp branch master).  Hopefully this one will be easier to merge.  If there are any other problems or if more input is needed from me, let me know.

### CODE REVIEW

File: src/signalrclient/json_helpers.cpp [hunk 1]
```diff
@@ -57,7 +58,42 @@ namespace signalr
             // because the server expects certain values to be 1 instead of 1.0 (like protocol version)
             if (std::modf(value, &intPart) == 0)
             {
-                return Json::Value((int)intPart);
+				if (value < 0)
```
Comment (halter73) [inline src/signalrclient/json_helpers.cpp]: Nit: Use spaces instead of tabs.

File: src/signalrclient/json_helpers.cpp [hunk 2]
```diff
@@ -57,7 +58,42 @@ namespace signalr
...
+					}
+					else if (value < INT64_MAX)
+					{
+						// Fits within int64_t
+						return Json::Value((int64_t)intPart);
+					}
+					else
+					{
+						// Needs uint64_t
+						return Json::Value((uint64_t)intPart);
```
Comment (halter73) [inline src/signalrclient/json_helpers.cpp]: Given that [doubles can contain exponents](https://en.wikipedia.org/wiki/IEEE_754), I'm pretty sure a non-fractional double can represent numbers larger than uint64_t's max value. The same issue exists for "large" negative numbers. For these values, we should just leave them as doubles.

**Comment by halter73** [comment]
Can we add some test cases? It would be nice to have at least one test case for each integral type we can transform the double into.

I'm also not sure it's worthwhile to support double, int32_t, int64_t, uint32_t, and uint64_t. I think we should just always try to convert non-fractional doubles to int64_t, and if that doesn't work just stick with double. That way consuming code doesn't need to be prepared to dealing with all those different integral types.

File: src/signalrclient/json_helpers.cpp (hunk 2 above)
Comment (willroman) [reply src/signalrclient/json_helpers.cpp]: @halter73, thanks for all of the comments.  This is feeling a bit like trial by fire for my first pull request but I'll keep working on it.

I agree with adding in logic to fall back to double for numbers outside of the range of the integer types.  I do have a question regarding supporting all of the integer types, however.  Given that in most cases the server side will be written in C#, does the SignalR server code take care of converting to the type that is given in the server-side methods?  If we were to implement this change to use just int64_t as you suggest, would the SignalR code handle the possible truncation if the server-side method specified its parameter as int or short?  Also, given that int64_t is signed, wouldn't that also open us back up to the overflow error I mentioned in #17779 (understanding that it would be far less likely to happen)?  If we use the smaller integer types on the C++ client, as I have it implemented now, would the SignalR server code handle the conversion if the server-side method specified its parameter as long or int? (I don't think it would be a problem but I figured I should ask)

I'm not sure how to go about adding in tests for this.  There doesn't seem to be an existing source file in the tests project that is really related to this.  Any suggestions you have in where to add the new tests would be greatly appreciated.

Comment (halter73) [reply src/signalrclient/json_helpers.cpp]: > Given that in most cases the server side will be written in C#, does the SignalR server code take care of converting to the type that is given in the server-side methods?

I think this is left up to the JSON serializer. Since JSON/JavaScript has only one "number" type which is stored in the IEEE 754 binary64 (i.e. float64_t or double\*) format, there's really no decision for the SignalR server to make regarding what type a number should be converted to.

It is possible that precision might be lost when converting an int64_t to a float64_t, but there should never be any truncation or overflow since the float64_t can use the exponent to represent values both much larger and much smaller than what can be represented using int64_t.

In general, SignalR does not use the payload itself to determine the type of a given number. On the server, if SignalR was expecting a double, it will deserialize the type as a double even if the value is is `0` which could easily be deserialized to a ushort or even a byte. The only reason the C++ client needs to basically guess the type of the number, is because c++ doesn't the powerful reflection as C# has.

I'm starting to think it might be simpler for the SignalR C++ client to treat all numbers as float64_t and let users do any conversion themselves if it's necessary.

> I'm not sure how to go about adding in tests for this. There doesn't seem to be an existing source file in the tests project that is really related to this. Any suggestions you have in where to add the new tests would be greatly appreciated.

@BrennanConroy Do you know of any similar tests?

\* In C++, a double might not actually be 64 bits

Comment (BrennanConroy) [reply src/signalrclient/json_helpers.cpp]: > On the server, if SignalR was expecting a double, it will deserialize the type as a double even if the value is is `0`

The "problem" is that with System.Text.Json as the default serializer now, it is a lot more strict about converting numbers between double and int, which is the whole reason the `std::modf` code exists.

We shied away from adding support for integral types to the `signalr::value` type until we needed it (like when adding MsgPack), but this conversation is making me think we might want to scrap the workaround and just add first party support for integral types.

Comment (BrennanConroy) [reply src/signalrclient/json_helpers.cpp]: > Do you know of any similar tests?

I would probably copy https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L410 and modify it for different argument types.

Comment (halter73) [reply src/signalrclient/json_helpers.cpp]: > We shied away from adding support for integral types to the signalr::value type until we needed it (like when adding MsgPack), but this conversation is making me think we might want to scrap the workaround and just add first party support for integral types.

The System.Text.Json thing is unfortunate, but I think we should stick with float64 as the only supported numeric type in the C++ client for not. Serializing as int64_t if possible in order to remove the decimal point and staying with float64_t when there's a fractional part or the number's too large seems to be the simplest fix for now. Hopefully we can get System.Text.Json to relax its behavior at some point.

File: src/signalrclient/json_helpers.cpp (hunk 1 above)
Comment (willroman) [reply src/signalrclient/json_helpers.cpp]: Converted all tabs to spaces in subsequent submissions.

File: src/signalrclient/json_helpers.cpp (hunk 2 above)
Comment (willroman) [reply src/signalrclient/json_helpers.cpp]: > > Do you know of any similar tests?
> 
> I would probably copy
> 
> https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L410
> 
> and modify it for different argument types.

Thanks for the suggestion.  To be forthcoming, C++ is not my primary language and I'm not familiar with the testing framework being used here.  I will try to find some time to develop a test method but it will take some time.

**Comment by BrennanConroy** [comment]
@willroman Have you had a chance to look at adding some tests (and fixing the tab formatting)?

**Comment by willroman** [comment]
@BrennanConroy:
> @willroman Have you had a chance to look at adding some tests (and fixing the tab formatting)?

I haven't had time to add tests yet but I fixed the tab formatting in my last commit.

**Comment by BrennanConroy** [comment]
Last commit didn't have any tab changes, but that's fine. Just wanted to ping the PR and see where we were at.

File: src/signalrclient/json_helpers.cpp (hunk 2 above)
Comment (willroman) [reply src/signalrclient/json_helpers.cpp]: > > Do you know of any similar tests?
> 
> I would probably copy
> 
> https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L410
> 
> and modify it for different argument types.

@BrennanConroy, I have been looking over the available test cases in the above and have not seen any examples of sending non-string payloads.  I also tried looking at the other tests but haven't found any non-string examples there either.  I'm not sure if I'm just missing them, however.  I am able to run the tests and have been experimenting with adding my own test cases but I could use some help with sending a non-string payload since the problem I'm trying to address is obviously related to that.  Thanks...

Comment (BrennanConroy) [reply src/signalrclient/json_helpers.cpp]: > have not seen any examples of sending non-string payloads

You're right, looks like I'll need to add some more tests later.

For now you could try looking at https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L796 and modifying it to send a non-empty array for args.

```c++
std::vector<signalr::value> arr { 100 };
hub_connection->invoke("method", signalr::value(arr), ...);
```

Comment (willroman) [reply src/signalrclient/json_helpers.cpp]: Added `TEST(invoke, invoke_handles_UINT32_MAX_value_as_unsigned_int)` test to validate that the issue I originally posted would now be handled correctly.  Probably still need 3-4 more tests to exercise all of the conditions added to the signalr::createJson() method in json_helpers.cpp.

File: test/signalrclienttests/hub_connection_impl_tests.cpp [hunk 3]
```diff
@@ -1164,3 +1164,83 @@ TEST(invoke, send_throws_when_the_underlying_connection_is_not_valid)
...
+
+TEST(invoke, invoke_handles_UINT32_MAX_value_as_unsigned_int)
+{
+    unsigned int original_value = 0xFFFFFFFF;
+
+    auto capture_message = false;
+    auto server_mre = manual_reset_event<std::string>();
+
+    std::ostringstream websocket_stream;
+    websocket_stream << "{ \"type\": 3, \"invocationId\": \"0\", \"result\": [" << original_value << "] }\x1e";
```
Comment (BrennanConroy) [inline test/signalrclienttests/hub_connection_impl_tests.cpp]: I think I'd prefer hard coding the string here

File: test/signalrclienttests/hub_connection_impl_tests.cpp [hunk 4]
```diff
@@ -1164,3 +1164,83 @@ TEST(invoke, send_throws_when_the_underlying_connection_is_not_valid)
...
+    std::string websocket_message = websocket_stream.str();
+
+    std::ostringstream expected_stream;
+    expected_stream << "{\"arguments\":[" << original_value << "],\"invocationId\":\"0\",\"target\":\"method\",\"type\":1}\x1e";
```
Comment (BrennanConroy) [inline test/signalrclienttests/hub_connection_impl_tests.cpp]: Same, hard code string

File: test/signalrclienttests/hub_connection_impl_tests.cpp [hunk 5]
```diff
@@ -1164,3 +1164,83 @@ TEST(invoke, send_throws_when_the_underlying_connection_is_not_valid)
...
+    std::string expected_message = expected_stream.str();
+
+    auto websocket_client = create_test_websocket_client(
+        /* send function */[&capture_message,&server_mre](const std::string& msg, std::function<void(std::exception_ptr)> callback)
+        {
+            if (capture_message)
```
Comment (BrennanConroy) [inline test/signalrclienttests/hub_connection_impl_tests.cpp]: I don't think you need this?

**Comment by BrennanConroy** [comment]
Oh, you need to rebase on master so this can build on the CI

**Comment by BrennanConroy** [comment]
Whoops, something went wrong. Did you do `git merge master` on the branch? I don't think that works, you need to `git rebase master` in order to get your changes applied on top of the master branch. If you reset to the previous commit I think you can then apply the rebase.

**Comment by willroman** [comment]
I think I made a real mess of things.  I'm not all that familiar with the Git command line (I do almost everything via Visual Studio).  Can you give me a hand in cleaning things up (reverting back to before the merge)?  I added @BrennanConroy and @halter73 as collaborators on my fork so you should, hopefully, be able to make the changes directly.  I managed to get the code back in shape locally but I can't figure out how to get things on the origin to be in that same state.

**Comment by BrennanConroy** [comment]
@willroman I moved your commits on top of master and added a commit to fix the compile errors in the test caused by some changes in the master branch.

**Comment by willroman** [comment]
@BrennanConroy Thanks so much for your help.  Sorry for making more work for you. I'm guessing that this branch can now cleanly merge with aspnet/master, correct?  Should I just add the other test cases now?

**Comment by BrennanConroy** [comment]
No worries 😄 

Yes this will merge cleanly now.

Please add the other tests!

File: src/signalrclient/json_helpers.cpp (hunk 2 above)
Comment (willroman) [reply src/signalrclient/json_helpers.cpp]: I've learned a lot more about the double (aka. float64) type in working on this PR.  All of the tests I have implemented currently pass but I'm beginning to rethink the edge cases because, as it turns out, double only has 52 bits to represent its mantissa (excluding the sign bit) so integral values contained within a double will be truncated if they exceed this.  I made the tests "work" by restricting the level of precision at which the values are compared.

This takes us back to the point that @BrennanConroy made on Dec 11, 2019: 

> > On the server, if SignalR was expecting a double, it will deserialize the type as a double even if the value is is `0`
> 
> The "problem" is that with System.Text.Json as the default serializer now, it is a lot more strict about converting numbers between double and int, which is the whole reason the `std::modf` code exists.
> 
> We shied away from adding support for integral types to the `signalr::value` type until we needed it (like when adding MsgPack), but this conversation is making me think we might want to scrap the workaround and just add first party support for integral types.

This PR, as implemented, will work for integral values that fit within 52 bits which gets us more support than before (i.e. it happens to support the case that I needed, a 32-bit mask ;-)) but it will fail for use cases outside of this range if full fidelity of the integral values are needed (e.g. using a 64-bit mask).  First party support for integral types would seem to be the only way to support those use cases.

**Comment by willroman** [comment]
@BrennanConroy, as I mentioned in a reply to one of the earlier questions, I have completed adding the test cases (a few days ago actually).   As I stated at that time..

> This PR, as implemented, will work for integral values that fit within 52 bits which gets us more support than before (i.e. it happens to support the case that I needed, a 32-bit mask ;-)) but it will fail for use cases outside of this range if full fidelity of the integral values are needed (e.g. using a 64-bit mask). First party support for integral types would seem to be the only way to support those use cases.

Are we good to go as is or would you like the test cases changed in any way?

File: test/signalrclienttests/hub_connection_tests.cpp [hunk 6]
```diff
@@ -22,6 +22,78 @@ hub_connection create_hub_connection(std::shared_ptr<websocket_client> websocket
...
+    ASSERT_FALSE(websocket_client->handshake_sent.wait(5000));
+
+    websocket_client->receive_message("{ }\x1e");
+
+    connect_mre.get();
+
+    auto callback_mre = manual_reset_event<signalr::value>();
+
+    std::vector<signalr::value> arr{ signalr::value((double)original_value) };
+    signalr::value args(arr);
```
Comment (BrennanConroy) [inline test/signalrclienttests/hub_connection_tests.cpp]: remove: unused
```suggestion
```

**Comment by BrennanConroy** [comment]
Thanks @willroman !
